### PR TITLE
Add Conditional Delete in frontend

### PR DESCRIPTION
**Pull request by jsjtzyy**
Add extra conditions to check for delete request in AmbrySecurityService.
If delete requestcontains Account name and Container name, frontend will
check if they match the Account and Container extracted from Blob Id.

### CODE REVIEW

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 1]
```diff
@@ -319,4 +329,39 @@ private void setAccountAndContainerHeaders(RestRequest restRequest, RestResponse
...
+   * @param restRequest the {@link RestRequest} that contains the {@link Account} and {@link Container} details.
+   * @throws RestServiceException if preconditions check failed.
+   */
+  private void checkForConditionalDelete(RestRequest restRequest) throws RestServiceException {
+    String accountNameFromHeader = getHeader(restRequest.getArgs(), Headers.TARGET_ACCOUNT_NAME, false);
+    String containerNameFromHeader = getHeader(restRequest.getArgs(), Headers.TARGET_CONTAINER_NAME, false);
+    if (accountNameFromHeader != null) {
+      Account targetAccount = (Account) restRequest.getArgs().get(InternalKeys.TARGET_ACCOUNT_KEY);
+      if (targetAccount == null) {
+        throw new RestServiceException("Account cannot be found", RestServiceErrorCode.InvalidAccount);
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: this is probably an internal server error

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 2]
```diff
@@ -319,4 +329,39 @@ private void setAccountAndContainerHeaders(RestRequest restRequest, RestResponse
...
+      String accountNameFromBlobId = targetAccount.getName();
+      if (!accountNameFromHeader.equals(accountNameFromBlobId)) {
+        throw new RestServiceException("Account name: " + accountNameFromHeader
+            + " from delete request doesn't match the account name from Blob id : " + accountNameFromBlobId,
+            RestServiceErrorCode.PreconditionFailed);
+      }
+      if (containerNameFromHeader != null) {
+        Container targetContainer = (Container) restRequest.getArgs().get(InternalKeys.TARGET_CONTAINER_KEY);
+        if (targetContainer == null) {
+          throw new RestServiceException("Container cannot be found", RestServiceErrorCode.InvalidContainer);
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: same here - internal server error

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 3]
```diff
@@ -319,4 +329,39 @@ private void setAccountAndContainerHeaders(RestRequest restRequest, RestResponse
...
+        throw new RestServiceException("Account cannot be found", RestServiceErrorCode.InvalidAccount);
+      }
+      String accountNameFromBlobId = targetAccount.getName();
+      if (!accountNameFromHeader.equals(accountNameFromBlobId)) {
+        throw new RestServiceException("Account name: " + accountNameFromHeader
+            + " from delete request doesn't match the account name from Blob id : " + accountNameFromBlobId,
+            RestServiceErrorCode.PreconditionFailed);
+      }
+      if (containerNameFromHeader != null) {
+        Container targetContainer = (Container) restRequest.getArgs().get(InternalKeys.TARGET_CONTAINER_KEY);
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: I actually have a utility for the check on line 344 and 355:
`RestUtils::getContainerFromArgs(...)` and `RestUtils::getAccountFromArgs(...)`

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 4]
```diff
@@ -319,4 +329,39 @@ private void setAccountAndContainerHeaders(RestRequest restRequest, RestResponse
...
+        if (targetContainer == null) {
+          throw new RestServiceException("Container cannot be found", RestServiceErrorCode.InvalidContainer);
+        }
+        String containerNameFromBlobId = targetContainer.getName();
+        if (!containerNameFromHeader.equals(containerNameFromBlobId)) {
+          throw new RestServiceException("Container name: " + containerNameFromHeader
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: One thing to keep in mind is that this check will only work for blobid v2+. I suppose we can explain that in client-facing docs.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 5]
```diff
@@ -319,4 +329,33 @@ private void setAccountAndContainerHeaders(RestRequest restRequest, RestResponse
...
     restResponseChannel.setHeader(RestUtils.Headers.PRIVATE, !container.isCacheable());
   }
+
+  /**
+   * Check precondtions for conditional delete if the {@code restRequest} contains the target
+   * account and container and they are not generic unknowns.
+   * @param restRequest the {@link RestRequest} that contains the {@link Account} and {@link Container} details.
+   * @throws RestServiceException if preconditions check failed.
+   */
+  private void checkForConditionalDelete(RestRequest restRequest) throws RestServiceException {
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: can you put this away as a `public static` function in `RestUtils`? - call it something like `accountAndContainerNamePreconditionCheck()`

File: ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java [hunk 6]
```diff
@@ -998,6 +1000,28 @@ private void setAmbryHeadersForPut(JSONObject headers, long ttlInSecs, boolean i
...
 
+  /**
+   * Sets headers for conditional delete.
+   * @param headers the {@link JSONObject} where the headers should be set.
+   * @param targetAccountName sets the {@link RestUtils.Headers#TARGET_ACCOUNT_NAME} header. Can be {@code null}.
+   * @param targetContainerName sets the {@link RestUtils.Headers#TARGET_CONTAINER_NAME} header. Can be {@code null}.
+   * @throws IllegalArgumentException if {@code headers} is null.
+   * @throws JSONException
+   */
+  private void setAmbryHeadersForDelete(JSONObject headers, String targetAccountName, String targetContainerName)
```
Comment (vgkholla) [inline ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java]: minor: call this `setAccountAndContainerHeaders()` - can be reused for non deletes also if required

File: ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java [hunk 7]
```diff
@@ -698,6 +698,34 @@ public static Container getContainerFromArgs(Map<String, Object> args) throws Re
...
+      if (containerNameFromHeader != null) {
+        Container targetContainer = getContainerFromArgs(restRequest.getArgs());
+        String containerNameFromBlobId = targetContainer.getName();
+        if (!containerNameFromHeader.equals(containerNameFromBlobId)) {
+          throw new RestServiceException("Container name: " + containerNameFromHeader
+              + "from request doesn't match the container name from Blob id : " + containerNameFromBlobId,
+              RestServiceErrorCode.PreconditionFailed);
+        }
+      }
+    }
```
Comment (cgtz) [inline ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java]: throw a badrequest exception if only the container name is set.

**Pull request merged**